### PR TITLE
feat: 符号化

### PR DESCRIPTION
**Pull request by maton369**

### CODE REVIEW

**Review by maton369 — COMMENTED** [review]
# GitHub Review — 4-13（BERT日本語・バッチ前処理／テンソル化）
対象コード：
tokenizer(
    text_list,
    max_length=10,
    padding='max_length',
    truncation=True,
    return_tensors='pt'
)

## 総評
実務でそのまま下流モデルへ渡せる B×L 形状（`input_ids`, `attention_mask`, `token_type_ids`）を生成しており、前処理の基本を正しく押さえている。特に `return_tensors='pt'` によるテンソル化は学習・推論の安定運用に必須である。本レビューでは理論面（Self-Attention とマスク、特殊トークンの不変量、計算量、再現性）を中心に改善提案を行う。

---

## 理論面の主要ポイント（確認事項）
1) 特殊トークンの不変量（単文）
- 先頭は `[CLS]`、有効長（non-PAD）の末尾は `[SEP]` が配置されるべきである。
- 有効長は `attention_mask.sum(dim=1)` に等しい。よって **`SEP` の位置 = 有効長 − 1** が成り立つことをアサートで検証できる。

2) Attention と PAD マスクの意味（学習への影響）
- Self-Attention は  
  **softmax(QKᵀ / √dₖ) V**。  
  `attention_mask=0` の位置はスコアへ **−∞** 相当が加算され、**注意が向かず勾配も基本的に流れない**。  
- 従って `padding='max_length'` を大きく取りすぎると、PADが増え **O(L²H)** の計算無駄が増える（Lは系列長）。固定長はスループット安定に利点があるが、**過剰PADは計算コスト**となる。

3) `token_type_ids` の位置づけ
- 単文は全0でよい（NSPやQAなど文対入力で A=0, B=1）。  
- 日本語BERT（Tohoku系）は `token_type_ids` を想定しているため、デフォルトの0列で整合する。

4) `truncation=True` の意味
- 単文では右側が切り詰められる（`[CLS]` は保持され、`[SEP]` は有効長末尾に再配置）。  
- 切り詰めは情報損失を生むため、**タスクの必要コンテキスト長**に応じて `max_length` を設計すること。

5) 再現性とデータ仕様
- 日本語では **形態素解析 → WordPiece** の二段で分割され、**辞書（ipadic / unidic-lite）と語彙・バージョン**によりトークン列・ID列は揺れる。  
- 再現性要件は「**性質（不変量）**」で置くのが実務的（例：特殊トークン位置、`attention_mask` 整合、最大長制約の順守）。

---

## 改善提案（理論準拠のアクション）
- ✅ **不変量アサートの追加**  
  単文前提で以下をテストに含めると安全性が上がる。  
  - `input_ids[:, 0] == cls_token_id`  
  - `sep_pos == attention_mask.sum(dim=1) - 1`（各サンプルで成り立つこと）  
  - 単文では `token_type_ids.eq(0).all()`  
- ✅ **動的パディングの併用**  
  学習時は `DataCollatorWithPadding`（`padding='longest'`）で **過剰PADを削減**し、O(L²H)の無駄を下げる。評価や導入フェーズでは固定長でスループット安定化、学習では動的化、などの切り分けを推奨。
- ✅ **長文タスクの設計**  
  QAや長文分類では `stride` と `return_overflowing_tokens=True` を利用し、スライディングウィンドウで文脈を保つ。右側切り詰め一発では解答スパンが消失しやすい。
- ✅ **ロス計算の整合**  
  トークン分類で PAD にラベルを付ける場合は、**`ignore_index`** を用意して PAD 位置の損失を無視するか、`attention_mask` で明示除外する。そうでないと **PADでロスが膨らむ**。
- ✅ **デバイス整合**  
  `batch_pt = {k: v.to(device) for k, v in batch_pt.items()}` で明示的にGPUへ移動。`model.to(device)` と整合させる。
- ✅ **入力正規化の一貫性**  
  CJK混在や全角/半角差への頑健性のため、必要に応じて `unicodedata.normalize("NFKC", text)` を前処理で適用（プロジェクト規約として固定）。

---

## 追加の理論観点（下流タスク別）
- 文分類：`[CLS]` 表現（`pooler_output` もしくは `last_hidden_state[:, 0, :]`）を線形ヘッドへ。固定長でも性能は安定しやすい。  
- トークン分類：サブワード境界とラベル拡張の整合が必須。`word_ids()`（fast系）やオフセット情報で語→サブワードへラベルを展開し、PADと分割サブワードを適切に無視する。  
- QA：スパンが `max_length` を跨ぎやすいので、**スライディング + ロジット統合**が定石。

---

## 小さな改善（メンテ性）
- ログ：`print` ではなく軽量ロガー（`logging`）で形状と不変量チェックを出す。  
- 設定の集中管理：`max_length` や `padding` 方針は設定ファイル／定数に寄せ、実験間での差分を追いやすくする。  
- 依存の固定：`transformers / fugashi / ipadic` のバージョンを `requirements.txt` へ固定し、CIで解決する。

---

## 結論
本セルは「固定長テンソルでBERTへ渡すための最小正攻法」を満たしている。理論的に重要なのは **(1) 特殊トークン位置の不変量**、**(2) PADマスクによるAttention抑制**、**(3) O(L²H)を意識した長さ設計** の3点である。これらをテストと前処理規約に落とすことで、下流学習・推論の安定性と再現性が大幅に向上する。
